KOGITO-8409 - Added info about CloudEvents to Knative functions (#312)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc

  ]
----

[[sending-cloudevents]]
==== Sending CloudEvents

By default, {product_name} sends the payload of a Knative function as a regular JSON object and `Content-Type` as `application/json`. However, you can tell {product_name} to send the payload as a CloudEvent. In that case, {product_name} will check if the CloudEvent has all mandatory attributes set and use `application/cloudevents+json; charset=UTF-8` in `Content-Type`.

To tell {product_name} you want to send the payload as a CloudEvent, you must define your function like:

[source,json]
----
  "functions": [
    {
      "name": "greet",
      "type": "custom",
      "operation": "knative:custom-function-knative-service",
      "metadata": {
        "path": "/",
        "asCloudEvent": true <1>
      }
    }
  ]
----

<1> Tells {product_name} to send the payload as a CloudEvent. Default is `false`

[WARNING]
====
You can't send a CloudEvent if you don't set the `asCloudEvent` attribute to `true`. If you do so, {product_name} will throw an error.
====

> 

Reviewer: This should be merged with the `Sending a CloudEvent` section. I don't see any benefit having it split in two places.

Author: Done

In case you need to send a payload in the request, you can add it to `arguments` in `functionRef`.

==== Sending a regular JSON object

> Given you want to send the following JSON object as the payload:

Reviewer 2: ```suggestion
You must send the following JSON object as the payload:
```

----
<1> The request payload is set in `arguments`.

[[sending-cloudevents]]
==== Sending a CloudEvent

By default, {product_name} sends the payload of a Knative function as a regular JSON object and `Content-Type` as `application/json`. However, you can tell {product_name} to send the payload as a CloudEvent. In that case, {product_name} will check if the CloudEvent has all mandatory attributes set and use `application/cloudevents+json; charset=UTF-8` in `Content-Type`.

> To tell {product_name} you want to send the payload as a CloudEvent, you must define your function like:

Reviewer 2: ```suggestion
To tell {product_name} you want to send the payload as a CloudEvent, you must define your function as follows:
```

[source,json]
----
  "functions": [
    {
      "name": "greet",
      "type": "custom",
      "operation": "knative:custom-function-knative-service",
      "metadata": {
        "path": "/",
        "asCloudEvent": true <1>
      }
    }
  ]
----

<1> Tells {product_name} to send the payload as a CloudEvent. Default is `false`

[WARNING]
====

> You can't send a CloudEvent if you don't set the `asCloudEvent` attribute to `true`. If you do so, {product_name} will throw an error.

Reviewer 2: ```suggestion
If you do not set the `asCloudEvent` attribute to `true`,  you can not send a CloudEvent. If you do so, {product_name} will throw an error.
```

====

Given you want to send the following CloudEvent as the payload:

[source,json]
----
{
  "specversion" : "1.0",
  "type" : "com.github.pull_request.opened",
  "source" : "https://github.com/cloudevents/spec/pull",
  "subject" : "123",
  "time" : "2018-04-05T17:31:00Z",
  "comexampleextension1" : "value",
  "comexampleothervalue" : 5,
  "datacontenttype" : "text/xml",
  "data" : "<much wow=\"xml\"/>"
}
----

> You must declare a `functionRef` like the following: (Don't forget to set `asCloudEvent` as `true` in function metadata)

Reviewer 2: ```suggestion
You must declare a `functionRef` like the following: (Do not forget to set `asCloudEvent` as `true` in function metadata)
```

[source,json]
----
  "states": [
    {
      "name": "invokeFunction",
      "type": "operation",
      "actions": [
        {
          "functionRef": {
            "refName": "greet",
            "arguments": { <1>
                "specversion" : "1.0",
                "type" : "com.github.pull_request.opened",
                "source" : "https://github.com/cloudevents/spec/pull",
                "subject" : "123",
                "time" : "2018-04-05T17:31:00Z",
                "comexampleextension1" : "value",
                "comexampleothervalue" : 5,
                "datacontenttype" : "text/xml",
                "data" : "<much wow=\"xml\"/>"
            }
          }
        }
      ],
      "end": true
    }
  ]
----
<1> The CloudEvent is set in `arguments`.

[NOTE]
====

> {product_name} generates a CloudEvent ID based on `source` and the workflow instance ID. In case an ID is set, {product_name} will ignore it and use a generated one.

Reviewer 2: ```suggestion
{product_name} generates a CloudEvent ID based on the `source` and the workflow instance ID. In case an ID is set, {product_name} will ignore it and use a generated one.
```

[WARNING]
====
You can't send a CloudEvent if you don't set the `asCloudEvent` attribute to `true`. If you do so, {product_name} will throw an error.
====

==== Sending a CloudEvent

Given you want to send the following CloudEvent as the payload:

[source,json]
----
{
  "specversion" : "1.0",
  "type" : "com.github.pull_request.opened",
  "source" : "https://github.com/cloudevents/spec/pull",
  "subject" : "123",
  "id" : "A234-1234-1234",
  "time" : "2018-04-05T17:31:00Z",
  "comexampleextension1" : "value",
  "comexampleothervalue" : 5,
  "datacontenttype" : "text/xml",
  "data" : "<much wow=\"xml\"/>"
}
----

You must declare a `functionRef` like the following: (Don't forget to set `asCloudEvent` as `true` in function metadata. See <<sending-cloudevents>>.)

[source,json]
----
  "states": [
    {
      "name": "invokeFunction",
      "type": "operation",
      "actions": [
        {
          "functionRef": {
            "refName": "greet",
            "arguments": { <1>
                "specversion" : "1.0",
                "type" : "com.github.pull_request.opened",
                "source" : "https://github.com/cloudevents/spec/pull",
                "subject" : "123",

>                 "id" : "A234-1234-1234",

Reviewer 3: Id can't be auto-generated by the caller (our engine)? It should. It's not a field users are expected to define in runtime.

Author: 1. If it's auto-generated, users won't be able to define a template. Something like `my_prefix{variable_id}`. Is that what we want?
2. If it's auto-generated, what will we use to generate it? Random, workflow_id, a template with a combination of properties...?

Reviewer 3: It's auto-generated, period.

You can open a JIRA to improve this. If it's not blank, verify if it's an expression. If so, evaluate it. If not, use it. Then you have the template thing if users need it.

But for this feature, as it is today, it should be generated. It's the easier way to be aligned with the spec for now: https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md#id

Author: Done.

----
<1> The request payload is set in `arguments`.

[WARNING]
====
You can't send a CloudEvent if you don't set the `asCloudEvent` attribute to `true`. If you do so, {product_name} will throw an error.

> ====

Reviewer: This is already mentioned.

Author: Done

----
<1> The request payload is set in `arguments`.

[WARNING]
====
You can't send a CloudEvent if you don't set the `asCloudEvent` attribute to `true`. If you do so, {product_name} will throw an error.
====

==== Sending a CloudEvent

Given you want to send the following CloudEvent as the payload:

[source,json]
----
{
  "specversion" : "1.0",
  "type" : "com.github.pull_request.opened",
  "source" : "https://github.com/cloudevents/spec/pull",
  "subject" : "123",

>   "id" : "A234-1234-1234",

Reviewer 3: ```suggestion

```
